# Analisis dan Pembersihan Data Fundamental Saham

## Pendahuluan

Dalam proyek ini, dilakukan analisis terhadap dataset fundamental saham yang memuat berbagai rasio dan metrik keuangan perusahaan. Data ini mencerminkan kondisi kesehatan keuangan perusahaan berdasarkan data historis dan proyeksi, yang sangat krusial dalam pengambilan keputusan investasi.

Dataset berisi beberapa indikator utama seperti:
- **Pendapatan per Saham (Revenue per Share)**
- **Price to Earnings Ratio (Trailing PE)**
- **EPS Trailing 12 Bulan**
- **EBITDA dan Total Revenue**
- **Debt to Equity Ratio**
- serta pertumbuhan pendapatan dan laba.

## Tujuan

Tujuan utama dari proyek ini adalah untuk melakukan _data wrangling_ dan _cleaning_ secara sistematis agar data siap dianalisis lebih lanjut, baik untuk eksplorasi data, visualisasi, hingga kemungkinan analisis lanjutan seperti segmentasi atau prediksi. Fokus khusus diberikan pada:
- Identifikasi dan penanganan **duplikasi dan nilai hilang**
- Deteksi dan perlakuan terhadap **outlier**
- Konversi satuan agar **konsisten dan mudah ditafsirkan**
- Perhitungan beberapa indikator turunan yang **bermanfaat dalam analisis fundamental**

Langkah-langkah ini penting untuk memastikan integritas dan validitas data yang akan digunakan sebagai dasar pengambilan keputusan analitik maupun finansial.

---


#Import Library

In [156]:
import pandas as pd
import numpy as np

#📥 Load datasets

In [157]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [158]:
stock_fundamental = pd.read_csv("drive/MyDrive/FUNDAMENTALratios.csv")
stock_fundamental

symbol  revenuePerShare  trailingPE  earningsQuarterlyGrowth  \
0       RELIANCE.BO         1296.823   28.012129                    0.093   
1       RELIANCE.NS         1296.823   28.010878                    0.093   
2       HDFCBANK.BO          293.740   16.260424                    0.359   
3     HINDUNILVR.NS          263.412   51.285230                    0.014   
4      ICICIBANK.NS          197.399   18.315136                    0.257   
...             ...              ...         ...                      ...   
5905     NIESSPM.BO              NaN         NaN                      NaN   
5906   QGOLDHALF.NS              NaN         NaN                      NaN   
5907  HDFCMFGETF.NS              NaN         NaN                      NaN   
5908  UTIBANKETF.NS              NaN   19.812199                      NaN   
5909   KOTAKGOLD.NS              NaN         NaN                      NaN   

      previousClose     open   dayLow  dayHigh      volume  trailingEps  \
0           2901.30  2897.05  2895.35  2920.00    562484.0       103.88   
1           2901.95  2899.95  2894.70  2920.00   9763420.0       103.89   
2           1445.10  1437.30  1437.30  1450.70    783356.0        88.74   
3           2242.35  2239.05  2232.05  2266.00   3507581.0        44.00   
4           1081.80  1081.15  1078.70  1093.70  17212189.0        59.53   
...             ...      ...      ...      ...         ...          ...   
5905           2.16     2.16     2.16     2.16       200.0          NaN   
5906          56.27    56.20    55.43    56.20     14972.0          NaN   
5907          51.67    51.81    51.51    51.81    385162.0          NaN   
5908          47.53    47.85    47.20    48.35    136993.0          NaN   
5909          57.02    56.64    56.22    56.70    416314.0          NaN   

      pegRatio        ebitda     totalDebt  totalRevenue  debtToEquity  \
0          NaN  1.503867e+12  3.166970e+12  8.773650e+12        36.100   
1          NaN  1.503867e+12  3.166970e+12  8.773650e+12        36.100   
2          NaN           NaN  7.996598e+12  1.933224e+12           NaN   
3          NaN  1.421525e+11  1.272000e+10  6.190100e+11         2.515   
4         1.03           NaN  2.009669e+12  1.380849e+12           NaN   
...        ...           ...           ...           ...           ...   
5905       NaN           NaN           NaN           NaN           NaN   
5906       NaN           NaN           NaN           NaN           NaN   
5907       NaN           NaN           NaN           NaN           NaN   
5908       NaN           NaN           NaN           NaN           NaN   
5909       NaN           NaN           NaN           NaN           NaN   

      revenuePerShare.1  earningsGrowth  revenueGrowth  
0              1296.823           0.093          0.036  
1              1296.823           0.093          0.036  
2               293.740          -0.001          1.211  
3               263.412           0.014         -0.002  
4               197.399           0.253          0.204  
...                 ...             ...            ...  
5905                NaN             NaN            NaN  
5906                NaN             NaN            NaN  
5907                NaN             NaN            NaN  
5908                NaN             NaN            NaN  
5909                NaN             NaN            NaN  

[5910 rows x 18 columns]

#🧱 STRUCTURING

In [159]:
stock_fundamental.info()
stock_fundamental.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5910 entries, 0 to 5909
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   symbol                   5910 non-null   object 
 1   revenuePerShare          3235 non-null   float64
 2   trailingPE               3382 non-null   float64
 3   earningsQuarterlyGrowth  2326 non-null   float64
 4   previousClose            5330 non-null   float64
 5   open                     5332 non-null   float64
 6   dayLow                   5332 non-null   float64
 7   dayHigh                  5332 non-null   float64
 8   volume                   5332 non-null   float64
 9   trailingEps              3242 non-null   float64
 10  pegRatio                 353 non-null    float64
 11  ebitda                   3012 non-null   float64
 12  totalDebt                3071 non-null   float64
 13  totalRevenue             3234 non-null   float64
 14  debtToEquity            

revenuePerShare   trailingPE  earningsQuarterlyGrowth  previousClose  \
count      3235.000000  3382.000000              2326.000000    5330.000000   
mean        368.680045          inf                 1.272781     667.639403   
std        1579.193992          NaN                11.447384    3202.109313   
min         -28.928000     0.025160                -0.998000       0.030000   
25%          44.690500    16.929313                -0.139000      29.650000   
50%         135.804000    29.994380                 0.187000     126.525000   
75%         351.737500    56.178136                 0.629000     484.075000   
max       58164.040000          inf               311.196000  131639.050000   

                open         dayLow        dayHigh        volume  trailingEps  \
count    5332.000000    5332.000000    5332.000000  5.332000e+03  3242.000000   
mean      668.518291     660.450069     682.882473  7.743663e+05    18.468205   
std      3202.498868    3175.467564    3241.002835  8.326325e+06   246.211146   
min         0.030000       0.030000       0.030000  1.000000e+00 -7848.510000   
25%        29.800000      29.087500      30.650000  2.978250e+03     0.522500   
50%       126.225000     124.000000     130.250000  1.891150e+04     8.535000   
75%       486.412500     479.662500     499.812500  1.242268e+05    27.590000   
max    131700.000000  130592.650000  132749.900000  4.684809e+08  4780.210000   

          pegRatio        ebitda     totalDebt  totalRevenue  debtToEquity  \
count   353.000000  3.012000e+03  3.071000e+03  3.234000e+03   2740.000000   
mean      7.752125  1.233124e+10  7.261417e+10  8.765737e+10    118.158228   
std      79.904143  6.869414e+10  4.394964e+11  4.458268e+11   1045.921778   
min    -120.870000 -2.968675e+10  1.000000e+05 -7.920600e+09      0.002000   
25%       0.520000  1.664015e+08  4.483450e+08  2.917744e+09      7.405500   
50%       1.260000  1.118691e+09  2.152000e+09  1.121537e+10     29.920000   
75%       2.770000  4.380904e+09  1.008934e+10  3.796004e+10     78.524500   
max    1059.900000  1.503867e+12  8.455658e+12  8.773650e+12  37076.363000   

       revenuePerShare.1  earningsGrowth  revenueGrowth  
count        3235.000000     2288.000000    3171.000000  
mean          368.680045        1.072592       0.490909  
std          1579.193992        8.258960       7.187965  
min           -28.928000       -0.998000      -0.999000  
25%            44.690500       -0.159000      -0.070500  
50%           135.804000        0.168000       0.069000  
75%           351.737500        0.592500       0.221000  
max         58164.040000      233.500000     191.250000

In [160]:
cols = [
    'symbol','revenuePerShare',            # pendapatan per saham
    'trailingPE',                 # price to earnings ratio
    'earningsQuarterlyGrowth',   # pertumbuhan laba per kuartal
    'trailingEps',               # EPS trailing 12 bulan
    'pegRatio',                  # PEG ratio
    'ebitda',                    # laba sebelum bunga dan penyusutan
    'totalDebt',                 # total utang
    'totalRevenue',              # total pendapatan
    'debtToEquity',              # rasio utang terhadap ekuitas
    'earningsGrowth',            # pertumbuhan laba tahunan
    'revenueGrowth'              # pertumbuhan pendapatan tahunan
]

stock_fundamental=stock_fundamental[cols]

In [161]:
stock_fundamental.rename(columns={
    'revenuePerShare': 'revenue_per_share',
    'trailingPE': 'trailing_pe',
    'earningsQuarterlyGrowth': 'earnings_quarterly_growth',
    'trailingEps': 'trailing_eps',
    'pegRatio': 'peg_ratio',
    'ebitda': 'ebitda',
    'totalDebt': 'total_debt',
    'totalRevenue': 'total_revenue',
    'debtToEquity': 'debt_to_equity',
    'earningsGrowth': 'earnings_growth',
    'revenueGrowth': 'revenue_growth'
}, inplace=True)

stock_fundamental.head()

<ipython-input-161-2901491438>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_fundamental.rename(columns={


symbol  revenue_per_share  trailing_pe  earnings_quarterly_growth  \
0    RELIANCE.BO           1296.823    28.012129                      0.093   
1    RELIANCE.NS           1296.823    28.010878                      0.093   
2    HDFCBANK.BO            293.740    16.260424                      0.359   
3  HINDUNILVR.NS            263.412    51.285230                      0.014   
4   ICICIBANK.NS            197.399    18.315136                      0.257   

   trailing_eps  peg_ratio        ebitda    total_debt  total_revenue  \
0        103.88        NaN  1.503867e+12  3.166970e+12   8.773650e+12   
1        103.89        NaN  1.503867e+12  3.166970e+12   8.773650e+12   
2         88.74        NaN           NaN  7.996598e+12   1.933224e+12   
3         44.00        NaN  1.421525e+11  1.272000e+10   6.190100e+11   
4         59.53       1.03           NaN  2.009669e+12   1.380849e+12   

   debt_to_equity  earnings_growth  revenue_growth  
0          36.100            0.093           0.036  
1          36.100            0.093           0.036  
2             NaN           -0.001           1.211  
3           2.515            0.014          -0.002  
4             NaN            0.253           0.204

In [162]:
#convert satuan ke persen
stock_fundamental.loc[:, 'earnings_quarterly_growth'] *= 100
stock_fundamental.loc[:, 'earnings_growth'] *= 100
stock_fundamental.loc[:, 'revenue_growth'] *= 100

stock_fundamental.head()

symbol  revenue_per_share  trailing_pe  earnings_quarterly_growth  \
0    RELIANCE.BO           1296.823    28.012129                        9.3   
1    RELIANCE.NS           1296.823    28.010878                        9.3   
2    HDFCBANK.BO            293.740    16.260424                       35.9   
3  HINDUNILVR.NS            263.412    51.285230                        1.4   
4   ICICIBANK.NS            197.399    18.315136                       25.7   

   trailing_eps  peg_ratio        ebitda    total_debt  total_revenue  \
0        103.88        NaN  1.503867e+12  3.166970e+12   8.773650e+12   
1        103.89        NaN  1.503867e+12  3.166970e+12   8.773650e+12   
2         88.74        NaN           NaN  7.996598e+12   1.933224e+12   
3         44.00        NaN  1.421525e+11  1.272000e+10   6.190100e+11   
4         59.53       1.03           NaN  2.009669e+12   1.380849e+12   

   debt_to_equity  earnings_growth  revenue_growth  
0          36.100              9.3             3.6  
1          36.100              9.3             3.6  
2             NaN             -0.1           121.1  
3           2.515              1.4            -0.2  
4             NaN             25.3            20.4

In [163]:
stock_fundamental.loc[:, 'ebitda'] = (stock_fundamental['ebitda'] / 1e9).round(2)
stock_fundamental.loc[:, 'total_debt'] = (stock_fundamental['total_debt'] / 1e9).round(2)
stock_fundamental.loc[:, 'total_revenue'] = (stock_fundamental['total_revenue'] / 1e9).round(2)
stock_fundamental.head()

symbol  revenue_per_share  trailing_pe  earnings_quarterly_growth  \
0    RELIANCE.BO           1296.823    28.012129                        9.3   
1    RELIANCE.NS           1296.823    28.010878                        9.3   
2    HDFCBANK.BO            293.740    16.260424                       35.9   
3  HINDUNILVR.NS            263.412    51.285230                        1.4   
4   ICICIBANK.NS            197.399    18.315136                       25.7   

   trailing_eps  peg_ratio   ebitda  total_debt  total_revenue  \
0        103.88        NaN  1503.87     3166.97        8773.65   
1        103.89        NaN  1503.87     3166.97        8773.65   
2         88.74        NaN      NaN     7996.60        1933.22   
3         44.00        NaN   142.15       12.72         619.01   
4         59.53       1.03      NaN     2009.67        1380.85   

   debt_to_equity  earnings_growth  revenue_growth  
0          36.100              9.3             3.6  
1          36.100              9.3             3.6  
2             NaN             -0.1           121.1  
3           2.515              1.4            -0.2  
4             NaN             25.3            20.4

In [164]:
stock_fundamental.rename(columns={
    'ebitda': 'ebitda_billionUSD',
    'total_debt': 'total_debt_billionUSD',
    'total_revenue': 'total_revenue_billionUSD'
}, inplace=True)
stock_fundamental.head()

<ipython-input-164-464041658>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_fundamental.rename(columns={


symbol  revenue_per_share  trailing_pe  earnings_quarterly_growth  \
0    RELIANCE.BO           1296.823    28.012129                        9.3   
1    RELIANCE.NS           1296.823    28.010878                        9.3   
2    HDFCBANK.BO            293.740    16.260424                       35.9   
3  HINDUNILVR.NS            263.412    51.285230                        1.4   
4   ICICIBANK.NS            197.399    18.315136                       25.7   

   trailing_eps  peg_ratio  ebitda_billionUSD  total_debt_billionUSD  \
0        103.88        NaN            1503.87                3166.97   
1        103.89        NaN            1503.87                3166.97   
2         88.74        NaN                NaN                7996.60   
3         44.00        NaN             142.15                  12.72   
4         59.53       1.03                NaN                2009.67   

   total_revenue_billionUSD  debt_to_equity  earnings_growth  revenue_growth  
0                   8773.65          36.100              9.3             3.6  
1                   8773.65          36.100              9.3             3.6  
2                   1933.22             NaN             -0.1           121.1  
3                    619.01           2.515              1.4            -0.2  
4                   1380.85             NaN             25.3            20.4

In [165]:
#cek jumlah duplikat
duplicate_rows = stock_fundamental.duplicated()
print("Jumlah duplikat:", duplicate_rows.sum())

Jumlah duplikat: 81


In [166]:
# Crosscheck apakah jumlah duplikat pada ticker sama dengan duplikat pada keseluruhan
ticker_counts = stock_fundamental['symbol'].value_counts()
duplicate_tickers = ticker_counts[ticker_counts > 1]

print("Jumlah ticker yang duplikat:", duplicate_tickers.shape[0])
print("Daftar ticker duplikat:")
print(duplicate_tickers)

Jumlah ticker yang duplikat: 81
Daftar ticker duplikat:
symbol
SRTRANSFIN-Z1.NS    2
SIDDHIKA-SM.NS      2
NIVAKA.BO           2
NIOSAD.BO           2
NBL.BO              2
                   ..
MEDPLUS.BO          2
ZMILGFIN.BO         2
ABSLBANETF.BO       2
M14RG-MF.NS         2
RITCO.BO            2
Name: count, Length: 81, dtype: int64


In [167]:
#karena sesuai jumlah dupilkat pada kode ticker dengan keseluruhan data, maka drop duplikat tersebut
stock_fundamental = stock_fundamental.drop_duplicates()

#Penanganan Pencilan

In [168]:
stock_fundamental.describe()

revenue_per_share  trailing_pe  earnings_quarterly_growth  \
count        3215.000000  3361.000000                2310.000000   
mean          368.925633          inf                 128.116537   
std          1583.680974          NaN                1148.644770   
min           -28.928000     0.025160                 -99.800000   
25%            44.690500    16.894735                 -13.700000   
50%           135.340000    29.959679                  18.700000   
75%           351.222000    56.178402                  62.975000   
max         58164.040000          inf               31119.600000   

       trailing_eps    peg_ratio  ebitda_billionUSD  total_debt_billionUSD  \
count   3222.000000   353.000000        2996.000000            3051.000000   
mean      18.434658     7.752125          12.385804              72.491980   
std      246.953671    79.904143          68.873123             440.358027   
min    -7848.510000  -120.870000         -29.690000               0.000000   
25%        0.500000     0.520000           0.170000               0.450000   
50%        8.535000     1.260000           1.110000               2.160000   
75%       27.590000     2.770000           4.380000              10.090000   
max     4780.210000  1059.900000        1503.870000            8455.660000   

       total_revenue_billionUSD  debt_to_equity  earnings_growth  \
count               3214.000000     2722.000000      2272.000000   
mean                  87.988460      118.325534       107.979357   
std                  447.175798     1049.295506       828.743746   
min                   -7.920000        0.002000       -99.800000   
25%                    2.900000        7.371000       -15.375000   
50%                   11.200000       29.920000        16.800000   
75%                   37.962500       78.493250        59.425000   
max                 8773.650000    37076.363000     23350.000000   

       revenue_growth  
count     3151.000000  
mean        49.343954  
std        721.065538  
min        -99.900000  
25%         -7.100000  
50%          6.900000  
75%         22.100000  
max      19125.000000

| Kolom                                     | Komentar                                                                                           |
| ----------------------------------------- | -------------------------------------------------------------------------------------------------- |
| `total_revenue_billionUSD` (max: 8773.65) | ✅ Masuk akal — bisa terjadi pada raksasa seperti Apple, Amazon, dll.                               |
| `total_debt_billionUSD` (max: 8455.66)    | ✅ Masuk akal, utang korporasi besar bisa triliunan.                                                |
| `ebitda_billionUSD` (max: 1503.87)        | ✅ Wajar, EBITDA besar pada perusahaan besar                                                        |
| `debt_to_equity` (max: 37076.36)          | ⚠️ Masih mungkin untuk industri highly-leveraged seperti banking, tapi perlu dicek siapa emitennya |
| `revenue_per_share` (max: 58164.04)       | ⚠️ Agak tinggi, tapi bisa terjadi pada perusahaan dengan very low outstanding shares               |


| Kolom                                       | Komentar                                                                                                                                                                              |
| ------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `trailing_pe` (max: inf)                    | ❌ Tidak masuk akal → seharusnya tidak boleh **inf**. Bisa disebabkan EPS = 0 atau missing value. Harus dihandle.                                                                      |
| `trailing_eps` (min: -7848.51)              | ⚠️ Perlu dicek. EPS negatif besar bisa jadi rugi besar (OK), tapi kalau hampir -8000 per lembar saham, sangat ekstrem, kemungkinan error parsing atau perusahaan tidak umum.          |
| `earnings_quarterly_growth` (max: 31119.6%) | ❌ Sangat tidak masuk akal. >31.000% artinya pertumbuhan laba 300x. Harus dicek datanya, kemungkinan salah input (misalnya 311.196 ditulis 31119.6)                                    |
| `earnings_growth` (max: 23350%)             | ❌ Kemungkinan sama seperti atas, angka tersebut ekstrem sekali, bahkan untuk startup.                                                                                                 |
| `revenue_growth` (max: 19125%)              | ❌ Terlalu ekstrem. Pertumbuhan revenue sebesar itu artinya revenue naik 192x. Perlu dicek apakah angka benar atau salah satuan.                                                       |
| `peg_ratio` (max: 1059.9, min: -120.87)     | ⚠️ Nilai normal PEG Ratio umumnya 0–3, >10 sudah langka. Nilai di atas 1000 dan negatif < -100 **perlu investigasi** lebih dalam. Kemungkinan data error atau hitungan tidak standar. |


In [169]:
# Ganti nilai inf pada trailing PE dengan NaN
stock_fundamental['trailing_pe'].replace([np.inf, -np.inf], np.nan, inplace=True)

# EPS terlalu besar atau kecil secara ekstrem
stock_fundamental['trailing_eps'] = stock_fundamental['trailing_eps'].mask(
    stock_fundamental['trailing_eps'].abs() > 1000
)

# PEG ratio tidak wajar (< -10 atau > 100)
stock_fundamental['peg_ratio'] = stock_fundamental['peg_ratio'].mask(
    (stock_fundamental['peg_ratio'] > 100) | (stock_fundamental['peg_ratio'] < -10)
)

# Debt to equity ratio ekstrem (> 1000)
stock_fundamental['debt_to_equity'] = stock_fundamental['debt_to_equity'].mask(
    stock_fundamental['debt_to_equity'] > 1000
)

# Growth metrics ekstrem (di luar -100% sampai 1000%)
growth_cols = ['earnings_growth', 'revenue_growth', 'earnings_quarterly_growth']
for col in growth_cols:
    stock_fundamental[col] = stock_fundamental[col].mask(
        (stock_fundamental[col] > 1000) | (stock_fundamental[col] < -100)
    )
stock_fundamental.describe()

<ipython-input-169-4127662677>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  stock_fundamental['trailing_pe'].replace([np.inf, -np.inf], np.nan, inplace=True)
<ipython-input-169-4127662677>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_fundamental['trailing_pe'].replace([np.inf, -np.inf], np.nan, inplace=True)
<ipython-input-169-41

revenue_per_share   trailing_pe  earnings_quarterly_growth  \
count        3215.000000   3354.000000                2265.000000   
mean          368.925633    130.857621                  43.249845   
std          1583.680974   1909.215146                 123.333667   
min           -28.928000      0.025160                 -99.800000   
25%            44.690500     16.847662                 -14.400000   
50%           135.340000     29.934276                  17.300000   
75%           351.222000     56.170623                  58.700000   
max         58164.040000  96462.500000                 941.100000   

       trailing_eps   peg_ratio  ebitda_billionUSD  total_debt_billionUSD  \
count   3214.000000  341.000000        2996.000000            3051.000000   
mean      22.254169    2.487889          12.385804              72.491980   
std       73.788100    5.896522          68.873123             440.358027   
min     -999.860000   -7.900000         -29.690000               0.000000   
25%        0.530000    0.570000           0.170000               0.450000   
50%        8.560000    1.300000           1.110000               2.160000   
75%       27.590000    2.770000           4.380000              10.090000   
max      787.660000   57.400000        1503.870000            8455.660000   

       total_revenue_billionUSD  debt_to_equity  earnings_growth  \
count               3214.000000     2689.000000      2227.000000   
mean                  87.988460       64.464691        39.674854   
std                  447.175798      105.392069       116.832331   
min                   -7.920000        0.002000       -99.800000   
25%                    2.900000        7.078000       -17.300000   
50%                   11.200000       29.120000        16.300000   
75%                   37.962500       76.352000        56.000000   
max                 8773.650000      897.369000       915.800000   

       revenue_growth  
count     3135.000000  
mean        12.784402  
std         61.415069  
min        -99.900000  
25%         -7.100000  
50%          6.700000  
75%         22.000000  
max        975.000000

## 🚧 Penanganan Outliers pada Data Fundamental Saham

Setelah melakukan eksplorasi awal terhadap statistik deskriptif, ditemukan bahwa beberapa kolom pada data memiliki nilai-nilai ekstrem (outliers) yang **secara logika bisnis tidak masuk akal atau sangat jarang terjadi**. Outliers semacam ini dapat mengganggu proses analisis dan menghasilkan insight yang tidak akurat, terutama jika digunakan dalam visualisasi atau pemodelan.

Berikut ini adalah beberapa penyesuaian yang dilakukan:

### 📌 1. Kolom `trailing_pe`
- Nilai `inf` atau `-inf` sering muncul karena pembagian dengan EPS nol atau sangat kecil.
- **Solusi:** Nilai `inf` dan `-inf` diganti dengan `NaN` agar dapat ditangani pada proses selanjutnya.

### 📌 2. Kolom `trailing_eps`
- Ditemukan nilai EPS yang sangat ekstrem (di atas 1000 atau di bawah -1000), yang kemungkinan merupakan kesalahan input atau kondisi luar biasa.
- **Solusi:** EPS dengan nilai absolut lebih dari 1000 diubah menjadi `NaN`.

### 📌 3. Kolom `peg_ratio`
- PEG ratio secara umum wajar di antara nilai positif kecil hingga sedang. Nilai di bawah -10 atau di atas 100 dianggap tidak wajar.
- **Solusi:** Diberikan batas bawah dan atas untuk menghapus nilai ekstrem tersebut.

### 📌 4. Kolom `debt_to_equity`
- Rasio ini bisa sangat tinggi untuk sektor tertentu, tetapi nilai di atas 1000 dianggap tidak wajar secara umum.
- **Solusi:** Nilai di atas 1000 diubah menjadi `NaN`.

### 📌 5. Kolom `growth` (earnings & revenue growth)
- Pertumbuhan lebih dari 1000% atau penurunan di bawah -100% dianggap terlalu ekstrem dan jarang ditemukan dalam praktik nyata.
- **Solusi:** Nilai-nilai tersebut disaring menjadi `NaN` karena dianggap outliers.

### ✅ Tujuan
Langkah ini bertujuan untuk meningkatkan kualitas data dan mencegah nilai-nilai ekstrem tersebut mendominasi hasil analisis. Outliers yang tidak relevan atau tidak realistis dapat menyebabkan distorsi pada metrik agregat seperti rata-rata, korelasi, atau mempengaruhi performa model jika digunakan untuk pembelajaran mesin.



#Checking Missing Values

In [170]:
stock_fundamental.isnull().sum()

symbol                          0
revenue_per_share            2614
trailing_pe                  2475
earnings_quarterly_growth    3564
trailing_eps                 2615
peg_ratio                    5488
ebitda_billionUSD            2833
total_debt_billionUSD        2778
total_revenue_billionUSD     2615
debt_to_equity               3140
earnings_growth              3602
revenue_growth               2694
dtype: int64

In [171]:
# Daftar kolom penting yang ingin dipastikan tidak ada NaN
kolom = [
    'revenue_per_share', 'trailing_pe', 'earnings_quarterly_growth',
    'trailing_eps', 'peg_ratio', 'ebitda_billionUSD', 'total_debt_billionUSD',
    'total_revenue_billionUSD', 'debt_to_equity', 'earnings_growth', 'revenue_growth'
]

# Drop baris yang punya missing value di kolom-kolom tersebut
stock_fundamental = stock_fundamental.dropna(subset=kolom).copy()

# Tampilkan jumlah baris setelah drop
print(f"Jumlah data setelah drop: {len(stock_fundamental)}")

Jumlah data setelah drop: 245


## 🔍 Penanganan Missing Values

Pada tahap pembersihan data (data cleaning), ditemukan cukup banyak nilai kosong (missing values) pada beberapa kolom fundamental seperti:

- `trailing_pe`
- `trailing_eps`
- `peg_ratio`
- `ebitda_billionUSD`
- `debt_to_equity`
- `earnings_growth`
- `revenue_growth`, dll.

Alih-alih melakukan imputasi (pengisian nilai kosong), diputuskan untuk **menghapus baris-baris yang memiliki nilai kosong** pada kolom-kolom penting tersebut. Keputusan ini didasarkan pada beberapa pertimbangan:

### 🧠 Alasan Menghapus Missing Values

1. **Berkaitan dengan data keuangan yang sensitif**
   > Nilai seperti `EPS`, `EBITDA`, atau `PEG Ratio` bersifat spesifik untuk masing-masing perusahaan. Mengisinya dengan rata-rata atau median bisa merusak makna asli data dan menyebabkan bias dalam analisis.

2. **Missing values kemungkinan besar tidak terjadi secara acak**
   > Banyak nilai yang hilang karena perusahaan belum mencetak laba, baru IPO, atau tidak relevan untuk sektor tertentu. Maka, imputasi justru bisa menyamarkan perbedaan penting antar perusahaan.

3. **Ukuran data tetap mencukupi**
   > Setelah pembersihan, jumlah data masih cukup besar untuk dianalisis tanpa kehilangan representasi.

### ✅ Kesimpulan

Maka, strategi yang digunakan adalah **drop baris dengan missing values** pada kolom-kolom fundamental utama agar kualitas data tetap terjaga dan analisis yang dilakukan bersifat valid serta dapat dipertanggungjawabkan.


#Constructing New Financial Indicators from Core Metrics

##PS Ratio

In [186]:
stock_fundamental['ps_ratio'] = stock_fundamental['trailing_pe'] * (stock_fundamental['trailing_eps'] / stock_fundamental['revenue_per_share'])

##Net Debt to EBITDA Ratio

In [187]:
stock_fundamental['net_debt_to_ebitda'] = stock_fundamental['total_debt_billionUSD'] / stock_fundamental['ebitda_billionUSD']

##Revenue to Debt Ratio

In [188]:
stock_fundamental['revenue_to_debt'] = stock_fundamental['total_revenue_billionUSD'] / stock_fundamental['total_debt_billionUSD']

##Earnings Yield (%)

In [189]:
stock_fundamental['earnings_yield_percent'] = (1 / stock_fundamental['trailing_pe']) * 100

##EBITDA Margin (%)

In [190]:
stock_fundamental['ebitda_margin_percent'] = (stock_fundamental['ebitda_billionUSD'] / stock_fundamental['total_revenue_billionUSD']) * 100

## 🔍 Membentuk Variabel Baru dari Rasio Fundamental Saham

Dalam analisis fundamental saham, data mentah seperti pendapatan, laba, dan utang seringkali belum cukup informatif. Oleh karena itu, diperlukan proses **rekayasa fitur (feature engineering)** untuk membentuk **rasio-rasio keuangan baru** yang dapat memberikan wawasan lebih dalam mengenai kesehatan finansial dan valuasi perusahaan.

Pada tahap ini, sejumlah variabel baru dihitung berdasarkan kombinasi dari kolom-kolom yang telah tersedia. Berikut adalah penjelasan setiap indikator yang dibentuk:

---

### 📌 1. Price to Sales Ratio (P/S Ratio)

**Definisi:**  
Rasio ini menunjukkan seberapa besar harga saham investor bersedia bayar untuk setiap satu dolar pendapatan yang dihasilkan perusahaan. Cocok untuk menilai perusahaan yang belum menghasilkan laba, namun memiliki potensi pertumbuhan pendapatan yang besar.

**Interpretasi:**  
- Nilai P/S tinggi → Saham dianggap mahal terhadap pendapatannya.  
- Nilai P/S rendah → Saham undervalued atau efisiensi pendapatan tinggi.


---

### 📌 2. Net Debt to EBITDA

**Definisi:**  
Mengukur kemampuan perusahaan dalam melunasi utangnya menggunakan pendapatan operasional sebelum bunga, pajak, depresiasi, dan amortisasi (EBITDA). Ini mencerminkan tingkat leverage perusahaan secara lebih realistik.

**Interpretasi:**  
- Nilai < 3 → Umumnya dianggap sehat.  
- Nilai tinggi → Perusahaan berisiko terhadap tekanan utang.

---

### 📌 3. Revenue to Debt Ratio

**Definisi:**  
Rasio ini menunjukkan berapa kali total pendapatan perusahaan dapat menutupi total utangnya. Sangat berguna untuk melihat kekuatan pendapatan terhadap beban kewajiban.

**Interpretasi:**  
- Nilai > 1 → Pendapatan lebih tinggi dari utang (positif).  
- Nilai < 1 → Utang melebihi pendapatan (berisiko).


---

### 📌 4. Earnings Yield (%)

**Definisi:**  
Menggambarkan return yang dihasilkan dari laba bersih terhadap harga saham. Ini adalah kebalikan dari Price to Earnings Ratio dan memungkinkan perbandingan dengan instrumen seperti obligasi.

**Interpretasi:**  
- Earnings Yield tinggi → Saham cenderung undervalued.  
- Earnings Yield rendah → Saham bisa jadi overvalued atau pertumbuhannya tinggi.


---

### 📌 5. EBITDA Margin (%)

**Definisi:**  
Menunjukkan seberapa besar laba operasional (EBITDA) yang dihasilkan dari setiap satu dolar pendapatan. Metrik ini menggambarkan efisiensi operasional perusahaan.

**Interpretasi:**  
- EBITDA Margin tinggi → Biaya operasional efisien, daya saing tinggi.  
- EBITDA Margin rendah → Perusahaan kurang efisien atau biaya tinggi.


---

### 📌 Kesimpulan

Pembuatan variabel-variabel baru ini bertujuan untuk membantu proses analisis data saham menjadi lebih terstruktur dan informatif. Dengan rasio-rasio tambahan ini, analis dapat menilai kondisi dan valuasi perusahaan dengan pendekatan yang lebih menyeluruh dan relevan terhadap kondisi pasar.


# 🧪 VALIDATION

In [177]:
# 1. Basic Info
print("Basic info:")
stock_fundamental.info()

Basic info:
<class 'pandas.core.frame.DataFrame'>
Index: 245 entries, 6 to 5664
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   symbol                     245 non-null    object 
 1   revenue_per_share          245 non-null    float64
 2   trailing_pe                245 non-null    float64
 3   earnings_quarterly_growth  245 non-null    float64
 4   trailing_eps               245 non-null    float64
 5   peg_ratio                  245 non-null    float64
 6   ebitda_billionUSD          245 non-null    float64
 7   total_debt_billionUSD      245 non-null    float64
 8   total_revenue_billionUSD   245 non-null    float64
 9   debt_to_equity             245 non-null    float64
 10  earnings_growth            245 non-null    float64
 11  revenue_growth             245 non-null    float64
 12  ps_ratio                   245 non-null    float64
 13  net_debt_to_ebitda         245 non-null   

In [178]:
# 2. Check for missing values
print("Missing values:")
print(stock_fundamental.isnull().sum())

Missing values:
symbol                       0
revenue_per_share            0
trailing_pe                  0
earnings_quarterly_growth    0
trailing_eps                 0
peg_ratio                    0
ebitda_billionUSD            0
total_debt_billionUSD        0
total_revenue_billionUSD     0
debt_to_equity               0
earnings_growth              0
revenue_growth               0
ps_ratio                     0
net_debt_to_ebitda           0
revenue_to_debt              0
earnings_yield_percent       0
ebitda_margin_percent        0
dtype: int64


In [179]:
# 3.cek duplikat
stock_fundamental.duplicated(subset='symbol').sum()

np.int64(0)

In [180]:
stock_fundamental.describe()

revenue_per_share  trailing_pe  earnings_quarterly_growth  \
count         245.000000   245.000000                 245.000000   
mean          577.276310    50.249424                  38.346939   
std           874.851099   107.347024                  98.564506   
min             4.481000     3.971563                 -95.700000   
25%           126.853000    20.587145                  -5.600000   
50%           301.862000    33.902650                  17.100000   
75%           636.543000    56.179580                  54.000000   
max          5595.403000  1635.500000                 672.100000   

       trailing_eps   peg_ratio  ebitda_billionUSD  total_debt_billionUSD  \
count    245.000000  245.000000         245.000000             245.000000   
mean      53.895184    2.439143          38.044122              75.475918   
std       86.674021    5.635574         103.843212             271.065741   
min        0.100000   -7.900000          -1.320000               0.000000   
25%       12.410000    0.650000           3.700000               1.250000   
50%       25.860000    1.470000           9.050000               4.960000   
75%       56.400000    2.900000          21.340000              20.230000   
max      621.470000   57.400000        1038.580000            2225.520000   

       total_revenue_billionUSD  debt_to_equity  earnings_growth  \
count                245.000000      245.000000       245.000000   
mean                 239.662286       37.450543        38.305306   
std                  672.521568       58.725025        98.633763   
min                    2.890000        0.009000       -95.800000   
25%                   28.140000        3.544000        -4.400000   
50%                   53.510000       14.980000        17.100000   
75%                  156.540000       46.343000        53.700000   
max                 5878.300000      605.424000       671.400000   

       revenue_growth    ps_ratio  net_debt_to_ebitda  revenue_to_debt  \
count      245.000000  245.000000          245.000000       245.000000   
mean         8.717143    7.406104            2.375383              inf   
std         19.840044   29.657679           12.166847              NaN   
min        -81.300000    0.157787           -2.250000         0.152421   
25%          0.500000    1.765292            0.150449         4.096490   
50%          8.000000    3.345495            0.724771        10.068828   
75%         15.300000    6.579269            1.744951        44.250000   
max        100.400000  336.721716          135.026316              inf   

       earnings_yield_percent  ebitda_margin_percent  
count              245.000000             245.000000  
mean                 3.858512              19.134486  
std                  3.407545              14.259154  
min                  0.061143             -29.139073  
25%                  1.780006              10.950413  
50%                  2.949622              15.950920  
75%                  4.857400              23.769554  
max                 25.179004              90.879924

In [181]:
# 5. Pastikan symbol benar-benar unik (jika diasumsikan satu baris = satu perusahaan)
stock_fundamental['symbol'].nunique() == stock_fundamental.shape[0]

True

In [182]:
# 6. Cek Nilai Negatif Tidak Masuk Akal
stock_fundamental[['ebitda_billionUSD', 'total_revenue_billionUSD', 'total_debt_billionUSD']].lt(0).sum()

ebitda_billionUSD           2
total_revenue_billionUSD    0
total_debt_billionUSD       0
dtype: int64

🔍 Analisis Validasi Nilai Negatif

- **EBITDA Negatif (2 baris)**  
  EBITDA yang bernilai negatif masih dianggap **masuk akal**, karena hal ini menunjukkan bahwa perusahaan mengalami kerugian dari aktivitas operasionalnya.  
  Kondisi ini bisa saja terjadi pada:
  - Perusahaan yang sedang dalam masa krisis keuangan
  - Perusahaan yang berada di fase turnaround
  - Sektor industri yang sangat kompetitif dengan margin rendah

- **Total Revenue dan Total Debt Negatif**
  - **Tidak ditemukan nilai negatif** pada kolom `total_revenue_billionUSD` maupun `total_debt_billionUSD`.
  - Ini adalah **hal yang baik**, karena secara logika keuangan, pendapatan dan total utang **tidak seharusnya bernilai negatif**.


In [183]:
stock_fundamental[['earnings_yield_percent', 'ps_ratio', 'net_debt_to_ebitda', 'revenue_to_debt', 'ebitda_margin_percent']].describe()

earnings_yield_percent    ps_ratio  net_debt_to_ebitda  \
count              245.000000  245.000000          245.000000   
mean                 3.858512    7.406104            2.375383   
std                  3.407545   29.657679           12.166847   
min                  0.061143    0.157787           -2.250000   
25%                  1.780006    1.765292            0.150449   
50%                  2.949622    3.345495            0.724771   
75%                  4.857400    6.579269            1.744951   
max                 25.179004  336.721716          135.026316   

       revenue_to_debt  ebitda_margin_percent  
count       245.000000             245.000000  
mean               inf              19.134486  
std                NaN              14.259154  
min           0.152421             -29.139073  
25%           4.096490              10.950413  
50%          10.068828              15.950920  
75%          44.250000              23.769554  
max                inf              90.879924

In [184]:
#ada nilai tidak masuk akal kembali muncul di var baru yaitu pada revenue to debt dan net debt to ebitda, berikut penanganannya
stock_fundamental['revenue_to_debt'].replace([np.inf, -np.inf], np.nan, inplace=True)
stock_fundamental = stock_fundamental[stock_fundamental['net_debt_to_ebitda'].between(0, 20)]

<ipython-input-184-1511480787>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  stock_fundamental['revenue_to_debt'].replace([np.inf, -np.inf], np.nan, inplace=True)


In [185]:
stock_fundamental[['earnings_yield_percent', 'ps_ratio', 'net_debt_to_ebitda', 'revenue_to_debt', 'ebitda_margin_percent']].describe()

earnings_yield_percent    ps_ratio  net_debt_to_ebitda  \
count              241.000000  241.000000          241.000000   
mean                 3.907154    7.272092            1.305480   
std                  3.414267   29.874898            1.609909   
min                  0.061143    0.157787            0.000000   
25%                  1.800040    1.731360            0.151381   
50%                  3.029215    3.293850            0.724771   
75%                  5.029881    6.315801            1.697368   
max                 25.179004  336.721716            8.884375   

       revenue_to_debt  ebitda_margin_percent  
count       238.000000             241.000000  
mean         68.981634              19.585196  
std         226.881037              13.837108  
min           0.152421               1.620765  
25%           4.110466              11.130868  
50%          10.068828              16.060413  
75%          41.323583              23.928660  
max        2664.000000              90.879924